streams/lms_metadata.py

```diff
@@ -0,0 +1,160 @@
+          self.server = f"{ip}:{port}"
+
+        if self.debug:
+          print(f"Found LMS Server: {ip}:{port}", flush=True)
+
+        player_json = {"id": 1,	"method": "slim.request",	"params": [self.player_name, ["players", "-", 100, "playerid"]]}
+        player_info = requests.get(f'http://{self.server}/jsonrpc.js', json=player_json, timeout=200)
+        player_load = json.loads(player_info.text)
+        players = player_load['result']['players_loop']
+
+        for player in players:
+          connected = player['connected']
+          if connected and player['name'] == self.player_name:
+            if self.debug:
+              print(f'LMS Metadata connected to player: {self.player_name}')
+            connected = True
+          else:
+            time.sleep(0.1)
+      except Exception as e:
+        # When first creating an LMS stream, there can be random errors that will close the while loop
+        # typically when asking the player for info when there isn't a player linked to the stream yet
+        if self.debug:
+          print(f"Could not find server: {e}", flush=True)
+        time.sleep(self.meta_ref_rate)
+
+    while connected:
+      try:
+        track_json = {"id": 1, "method": "slim.request", "params": [ self.player_name, ["status", "-",100] ]}
+        track_info = requests.post(f'http://{self.server}/jsonrpc.js 2>/dev/null', json=track_json, timeout=200)
+        track_load = json.loads(track_info.text)
+
+        track_id = track_load["result"]["playlist_loop"][0]["id"]
+        song_json = {"id":2,"method":"slim.request","params":[ self.player_name, ["songinfo","-",100,f"track_id:{track_id}"]]}
+        song_info = requests.post(f'http://{self.server}/jsonrpc.js 2>/dev/null', json=song_json, timeout=200)
+        song_load = json.loads(song_info.text)
+
+        song_data = self.flatten(song_load['result']['songinfo_loop'])
+        track_data = self.flatten(track_load['result']['playlist_loop'])
+
+        meta = {
+          'artist': 'Loading...',
+          'album': 'If loading takes a long time,',
+          'track': 'consider adding hostname to stream config',
+          'image_url': 'static/imgs/lms.png'
+         }
+
+        if self.debug:
+          print(f"Current stream type: {song_data['type']}")
+
+        if song_data['type'] == "MP3 Radio" or song_data['type'] == "AAC Radio" or song_data['type'] == "Radio" or song_data['type'] == "TEXT/X-JSON Radio":
+          try:
+            meta["track"] = track_data["title"]
+            meta['artist'] = None
+            meta['album'] = song_data['remote_title']
+            meta["image_url"] = f"http://{self.server}/music/{song_data['coverid']}/cover.jpg?id={song_data['coverid']}"
+          except KeyError:
+            # Sometimes, KeyError will occur when switching between streams with different metadata formatting
+            print(f"KeyError, trying again in {self.meta_ref_rate} seconds...")
+
+            meta["track"] = song_data["title"]
```

Comment:
I would suggest fully updating meta in the case you don't update all of the fields to avoid having lingering old values intermixed with new data. You could use a data class to handle this .